### PR TITLE
feat(subscribe): WIP for subscription

### PR DESCRIPTION
**Pull request by vringar**
Co-authored-by: TuCl <75214650+TuCl@users.noreply.github.com>
Co-authored-by: Emirhan Atabay <e.atabay3@googlemail.com>
Co-authored-by: Jan Golebiowski <golebiowski@campus.tu-berlin.de>

# Pull Request Checklist

## Author

- [ ] Built code and tested locally
- [ ] Removed any TODOs, etc.
- [ ] Metioned the issue this PR closes
- [ ] Good description of what the PR does written

## Reviewer

- [ ] Built and tested code locally
- [ ] Fully understood the PR
- [ ] All open conversations resolved

### CODE REVIEW

**Comment by vringar** [comment]
@Ti-ger let's review and merge this whenever you're available.

File: src/main/SubscriptionManager.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace chip {
+SubscriptionManager SubscriptionManager::sSubscriptionManager;
+CHIP_ERROR SubscriptionManager::RegisterSubscription(shell::SubscribeCommandData * data)
+{
+    // Send subscribe request
+    SubscriptionManager::SendSubscribeRequest(data);
+    // Receive report manager?!
+
+    // send status response ?!
```
Comment (vringar) [inline src/main/SubscriptionManager.cpp:20]: Do we know if we need to do more here?

File: src/main/SubscriptionManager.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        ChipLogError(NotSpecified, "ClusterCommandWorkerFunction - Unable to find the mentioned Peer");
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    ESP_LOGI("Subscribe", "SubscribeSubscribeCommandHandler");
+    ESP_LOGI("Subscribe", " - EndPoint ID: '0x%02x'", data->endpointId);
+    ESP_LOGI("Subscribe", " - Cluster ID: '0x%02x'", data->clusterId);
+    ESP_LOGI("Subscribe", " - Attribute ID: '0x%02x'", data->attributeId);
+    ESP_LOGI("Subscribe", " - minInterval: '0x%02x'", data->minInterval);
+    ESP_LOGI("Subscribe", " - maxInterval: '0x%02x'", data->maxInterval);
```
Comment (vringar) [inline src/main/SubscriptionManager.cpp:50]:
```suggestion
```

File: src/main/include/Subscription.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            mError = error;
+            return;
+        }
+
+        if (data == nullptr)
+        {
+            ChipLogError(chipTool, "Response Failure: No Data");
+            mError = CHIP_ERROR_INTERNAL;
+            return;
+        }
```
Comment (vringar) [inline src/main/include/Subscription.h:55]: Should we do more here or should this be reserved for #43?

File: src/main/include/SubscriptionManager.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    SubscriptionManager()
+    {
+        auto & server       = chip::Server::GetInstance();
+        mFabricTable        = &server.GetFabricTable();
+        mCASESessionManager = server.GetCASESessionManager();
+    };
+    /*CHIP_ERROR RegisterSubscription(chip::FabricId fabricId, chip::app::ConcreteAttributePath attributesPath,
+                                    chip::SubscriptionId * subscriptionId); // subscriptionId is out param
+    CHIP_ERROR RegisterSubscription(chip::FabricId fabricId, chip::app::ConcreteEventPath eventPath,
+                                    chip::SubscriptionId * subscriptionId); // subscriptionId is out param*/
```
Comment (vringar) [inline src/main/include/SubscriptionManager.h:31]:
```suggestion
```

File: src/main/SubscriptionManager.cpp (hunk 1 above)
Comment (Ti-ger) [reply src/main/SubscriptionManager.cpp:20]: I don't think so. 

The interaction model already contains the required callbacks

File: src/main/include/Subscription.h (hunk 3 above)
Comment (Ti-ger) [reply src/main/include/Subscription.h:55]: Let's do this in its own PR. This one is already quite bloated

**Review by Ti-ger — APPROVED** [review]

**Pull request merged**